invoice-api/src/service/freshbooks.ts

```diff
@@ -0,0 +1,503 @@
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
+                return response.data.response.result;
+            }).catch(err => {
+                throw new Error(err.response.statusText);
+            });
+    }
+    private getCustomerIDs(customersInfo:any): any{
+        let output = [];
+        customersInfo.forEach((customerInfo) => {output.push(customerInfo.id)}, customersInfo);
+        return output;
+    }
+
+    public async emailExistingInvoice(customerEmail:string, invoiceNumber: string): Promise<any> {
+        await this.authenticate();
+        let invoiceInfo = await this.findInvoice(invoiceNumber);
+        let customerInfo = await this.getCustomerInfo(invoiceInfo.customerid);
+        // email if customerEmail exists in one of the contacts
+        if(!this.validEmailForCustomer(customerEmail,customerInfo))
+            throw Error("error: User account: "+ customerEmail +" does not have access to this Invoice: "+invoiceNumber );
+        let emailInfoJson = this.getExistingInvoiceEmailInfo(customerEmail,invoiceNumber);
+        return this.emailInvoice(invoiceInfo.id,emailInfoJson);
+    }
+
+    private validEmailForCustomer(email:string, customerInfo:any): boolean{
+        let lowerCasedEmail = email.toLowerCase();
+        if(customerInfo.email != ''){
+            if(customerInfo.email.toLowerCase() == lowerCasedEmail) return true;
+        }
+        // seach for email in secondary contacts
+        let contact:any;
+        for(let contactIdx in customerInfo.recent_contacts){
+            contact = customerInfo.recent_contacts[contactIdx];
+            if(contact.email != ''){
+                if(contact.email.toLowerCase() == lowerCasedEmail) return true;
+            }
+        } // contacts iteration ends here
+        return false;
+    }
+
+    private async findInvoice(invoiceNumber:string): Promise<any> {
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/invoices/invoices?search[invoice_number]=${invoiceNumber}`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
+                return response.data.response.result.invoices[0];
+            }).catch(err => {
+                throw new Error(err.response.statusText);
+            })
+    }
+
+    private async getCustomerInfo(customerId:string){
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/users/clients?search[userid]=${customerId}&include[]=recent_contacts`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```

Comment:
another one here, will probably end up being too verbose. Also without timestamps reducing the usefulness.